File: packages/deno/src/expr.ts
// ---------------------------------------------------------------------------
// Expression evaluator — supports dot access, bracket access, function calls,
// ternary, string concatenation, comparisons, and unary !.
// ---------------------------------------------------------------------------

type Tok =
  | { t: "n"; v: number }
  | { t: "s"; v: string }
  | { t: "i"; v: string }
  | { t: "o"; v: string }
  | { t: "r"; v: RegExp };

function tokenize(src: string): Tok[] {
  const out: Tok[] = [];
  let i = 0;
  while (i < src.length) {
    if (/\s/.test(src[i])) { i++; continue; }
    // numbers
    if (/\d/.test(src[i])) {
      let n = "";
      while (i < src.length && /[\d.]/.test(src[i])) n += src[i++];
      out.push({ t: "n", v: Number(n) });
      continue;
    }
    // strings
    if (src[i] === '"' || src[i] === "'") {
      const q = src[i++];
      let s = "";
      while (i < src.length && src[i] !== q) {
        if (src[i] === "\\") { i++; s += src[i++]; continue; }
        s += src[i++];
      }
      i++; // closing quote
      out.push({ t: "s", v: s });
      continue;
    }
    // regex literals: /pattern/flags — only after operator or at start
    if (src[i] === '/' && (out.length === 0 || out[out.length - 1].t === "o" || (out[out.length - 1].t === "o" && out[out.length - 1].v === ","))) {
      const start = i;
      i++; // skip opening /
      let pattern = "";
      while (i < src.length && src[i] !== '/') {
        if (src[i] === '\\') { pattern += src[i++]; if (i < src.length) pattern += src[i++]; continue; }
        pattern += src[i++];
      }
      i++; // skip closing /
      let flags = "";
      while (i < src.length && /[gimsuvy]/.test(src[i])) flags += src[i++];
      try {
        out.push({ t: "r", v: new RegExp(pattern, flags) });
      } catch {
        // Fall back to treating / as operator
        i = start;
        out.push({ t: "o", v: src[i++] });
      }
      continue;
    }
    // identifiers
    if (/[a-zA-Z_$]/.test(src[i])) {
      let id = "";
      while (i < src.length && /[a-zA-Z0-9_$]/.test(src[i])) id += src[i++];
      out.push({ t: "i", v: id });
      continue;
    }
    // multi-char ops
    const tri = src.substring(i, i + 3);
    if (tri === "===" || tri === "!==") { out.push({ t: "o", v: tri }); i += 3; continue; }
    const bi = src.substring(i, i + 2);
    if (bi === "==" || bi === "!=" || bi === ">=" || bi === "<=") {
      out.push({ t: "o", v: bi }); i += 2; continue;
    }
    // single-char op
    out.push({ t: "o", v: src[i++] });
  }
  return out;
}

export function evalExpr(src: string, ctx: Record<string, unknown>): unknown {
  const toks = tokenize(src);
  let pos = 0;
  const peek = (): Tok | undefined => toks[pos];
  const adv = (): Tok => toks[pos++];
  const matchOp = (v: string): boolean => {
    if (peek()?.t === "o" && peek()?.v === v) { pos++; return true; }
    return false;
  };

  function ternary(): unknown {
    const lhs = addition();
    if (matchOp("?")) { const a = addition(); matchOp(":"); const b = addition(); return lhs ? a : b; }
    return lhs;
  }

  function addition(): unknown {
    let left = comparison();
    while (peek()?.t === "o" && peek()?.v === "+") {
      adv();
      const right = comparison();
      if (typeof left === "number" && typeof right === "number") left = left + right;
      else left = String(left ?? "") + String(right ?? "");
    }
    return left;
  }

  function comparison(): unknown {
    const left = unary();
    const t = peek();
    if (t?.t === "o" && ["===", "!==", "==", "!=", ">", "<", ">=", "<="].includes(t.v)) {
      adv();
      const right = unary();
      switch (t.v) {
        case "===": return left === right;
        case "!==": return left !== right;
        case "==":  return left == right;
        case "!=":  return left != right;
        case ">":   return (left as number) > (right as number);
        case "<":   return (left as number) < (right as number);
        case ">=":  return (left as number) >= (right as number);
        case "<=":  return (left as number) <= (right as number);
      }
    }
    return left;
  }

  function unary(): unknown {
    if (matchOp("!")) return !unary();
    return postfix();
  }

  function postfix(): unknown {
    let val = primary();
    while (true) {
      if (matchOp(".")) {
        const id = adv();
        if (val != null) {
          const prop = (val as Record<string, unknown>)[id.v as string];
          // Bind methods to their owner so that e.g. "str".endsWith('.pdf') works
          val = typeof prop === "function" ? (prop as CallableFunction).bind(val) : prop;
        } else {
          val = undefined;
        }
      } else if (matchOp("[")) {
        const idx = ternary();
        matchOp("]");
        val = val != null ? (val as Record<string, unknown>)[String(idx)] : undefined;
      } else if (matchOp("(")) {
        const args: unknown[] = [];
        if (!matchOp(")")) {
          args.push(ternary());
          while (matchOp(",")) args.push(ternary());
          matchOp(")");
        }
        val = typeof val === "function" ? val(...args) : undefined;
      } else break;
    }
    return val;
  }

  function primary(): unknown {
    const t = peek();
    if (!t) return undefined;
    if (t.t === "n") { adv(); return t.v; }
    if (t.t === "s") { adv(); return t.v; }
    if (t.t === "r") { adv(); return t.v; }
    if (t.t === "i") {
      adv();
      if (t.v === "true") return true;
      if (t.v === "false") return false;
      if (t.v === "null") return null;
      if (t.v === "undefined") return undefined;
      return ctx[t.v];
    }
    if (t.t === "o" && t.v === "(") { adv(); const v = ternary(); matchOp(")"); return v; }
    // Array literal: [expr, expr, ...]
    if (t.t === "o" && t.v === "[") {
      adv();
      const arr: unknown[] = [];
      if (!matchOp("]")) {
        arr.push(ternary());
        while (matchOp(",")) arr.push(ternary());
        matchOp("]");
      }
      return arr;
    }
    // Object literal: {key: value, ...}
    if (t.t === "o" && t.v === "{") {
      adv();
      const obj: Record<string, unknown> = {};
      if (!matchOp("}")) {
        const k = adv();
        matchOp(":");
        obj[String(k.v)] = ternary();
        while (matchOp(",")) {
          const k2 = adv();
          matchOp(":");
          obj[String(k2.v)] = ternary();
        }
        matchOp("}");
      }
      return obj;
    }
    return undefined;
  }

  return ternary();
}

export function interpolate(text: string, ctx: Record<string, unknown>): string {
  if (!text.includes("{{")) return text;

  // Detect if trailing content after last {{ }} is whitespace-only
  const lastClose = text.lastIndexOf("}}");
  const trimTail =
    lastClose !== -1 && /^\s*$/.test(text.substring(lastClose + 2));

  let out = text.replace(/\{\{(.+?)\}\}/g, (_, e: string) => {
    const v = evalExpr(e.trim(), ctx);
    if (v == null) return "";
    // Booleans render as empty (matching React behavior)
    if (typeof v === "boolean") return "";
    return String(v);
  });

  if (trimTail) out = out.replace(/\s+$/, "");
  return out;
}

/** Interpolation without trailing whitespace trimming — for pre mode. */
export function interpolatePre(text: string, ctx: Record<string, unknown>): string {
  if (!text.includes("{{")) return text;
  return text.replace(/\{\{(.+?)\}\}/g, (_, e: string) => {
    const v = evalExpr(e.trim(), ctx);
    if (v == null) return "";
    if (typeof v === "boolean") return "";
    return String(v);
  });
}

export function evalCondition(expr: string, ctx: Record<string, unknown>): boolean {
  const m = expr.match(/^\{\{(.+?)\}\}$/);
  if (m) return !!evalExpr(m[1].trim(), ctx);
  return !!evalExpr(expr, ctx);
}

export function escapes(text: string): string {
  return text
    // POML escape sequences
    .replace(/#quot;/g, '"')
    .replace(/#apos;/g, "'")
    .replace(/#lt;/g, "<")
    .replace(/#gt;/g, ">")
    .replace(/#lbrace;/g, "{")
    .replace(/#rbrace;/g, "}")
    .replace(/#amp;/g, "\x00")
    .replace(/#hash;/g, "#")
    // XML entities
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&amp;/g, "")
    // deno-lint-ignore no-control-regex
    .replace(/\x00/g, "&");
}

File: packages/python/pomlight/expr.py
from __future__ import annotations

import re
from typing import Any


def _tokenize(src: str) -> list[dict]:
    out: list[dict] = []
    i = 0
    while i < len(src):
        ch = src[i]
        # whitespace
        if ch in " \t\r\n":
            i += 1
            continue
        # numbers
        if ch.isdigit():
            n = ""
            while i < len(src) and src[i] in "0123456789.":
                n += src[i]
                i += 1
            out.append({"t": "n", "v": float(n) if "." in n else int(n)})
            continue
        # strings
        if ch in ('"', "'"):
            q = ch
            i += 1
            s = ""
            while i < len(src) and src[i] != q:
                if src[i] == "\\":
                    i += 1
                    if i < len(src):
                        s += src[i]
                        i += 1
                    continue
                s += src[i]
                i += 1
            i += 1  # closing quote
            out.append({"t": "s", "v": s})
            continue
        # identifiers
        if ch.isalpha() or ch in "_$":
            ident = ""
            while i < len(src) and (src[i].isalnum() or src[i] in "_$"):
                ident += src[i]
                i += 1
            out.append({"t": "i", "v": ident})
            continue
        # regex literals: /pattern/flags — only after operator or at start
        if ch == "/" and (len(out) == 0 or out[-1]["t"] == "o"):
            start = i
            i += 1  # skip opening /
            pattern = ""
            while i < len(src) and src[i] != "/":
                if src[i] == "\\":
                    pattern += src[i]
                    i += 1
                    if i < len(src):
                        pattern += src[i]
                        i += 1
                    continue
                pattern += src[i]
                i += 1
            i += 1  # skip closing /
            flags = ""
            while i < len(src) and src[i] in "gimsuvy":
                flags += src[i]
                i += 1
            # Convert JS flags to Python: g → global (handled by re.sub), m → re.MULTILINE, i → re.IGNORECASE
            py_flags = 0
            is_global = False
            for f in flags:
                if f == "g":
                    is_global = True
                elif f == "i":
                    py_flags |= re.IGNORECASE
                elif f == "m":
                    py_flags |= re.MULTILINE
                elif f == "s":
                    py_flags |= re.DOTALL
            try:
                compiled = re.compile(pattern, py_flags)
                out.append({"t": "r", "v": _JsRegex(compiled, is_global)})
            except re.error:
                # Fallback: treat / as operator
                i = start
                out.append({"t": "o", "v": src[i]})
                i += 1
            continue
        # multi-char ops
        tri = src[i : i + 3]
        if tri in ("===", "!=="):
            out.append({"t": "o", "v": tri})
            i += 3
            continue
        bi = src[i : i + 2]
        if bi in ("==", "!=", ">=", "<="):
            out.append({"t": "o", "v": bi})
            i += 2
            continue
        # single-char op
        out.append({"t": "o", "v": ch})
        i += 1
    return out


class _JsRegex:
    """Wrapper for a compiled regex with JS global flag tracking."""
    __slots__ = ("pattern", "is_global")
    def __init__(self, pattern: re.Pattern, is_global: bool):
        self.pattern = pattern
        self.is_global = is_global
    def sub(self, repl: str, s: str) -> str:
        if self.is_global:
            return self.pattern.sub(repl, s)
        return self.pattern.sub(repl, s, count=1)


def _js_replace(s: str, pattern: Any, replacement: str, replace_all: bool) -> str:
    """Handle JS-style string replace with regex or string patterns."""
    if isinstance(pattern, _JsRegex):
        return pattern.sub(replacement, s)
    if hasattr(pattern, "sub"):
        # compiled re.Pattern object
        return pattern.sub(replacement, s)
    # String pattern
    if replace_all:
        return s.replace(str(pattern), replacement)
    return s.replace(str(pattern), replacement, 1)


def eval_expr(src: str, ctx: dict[str, Any]) -> Any:
    toks = _tokenize(src)
    pos = [0]

    def peek() -> dict | None:
        return toks[pos[0]] if pos[0] < len(toks) else None

    def adv() -> dict:
        t = toks[pos[0]]
        pos[0] += 1
        return t

    def match_op(v: str) -> bool:
        p = peek()
        if p and p["t"] == "o" and p["v"] == v:
            pos[0] += 1
            return True
        return False

    def ternary() -> Any:
        lhs = addition()
        if match_op("?"):
            a = addition()
            match_op(":")
            b = addition()
            return a if lhs else b
        return lhs

    def addition() -> Any:
        left = comparison()
        while True:
            p = peek()
            if p and p["t"] == "o" and p["v"] == "+":
                adv()
                right = comparison()
                if isinstance(left, (int, float)) and isinstance(right, (int, float)):
                    left = left + right
                else:
                    left = str(left if left is not None else "") + str(right if right is not None else "")
            else:
                break
        return left

    def comparison() -> Any:
        left = unary()
        p = peek()
        if p and p["t"] == "o" and p["v"] in ("===", "!==", "==", "!=", ">", "<", ">=", "<="):
            adv()
            right = unary()
            op = p["v"]
            if op in ("===", "=="):
                return left == right
            if op in ("!==", "!="):
                return left != right
            if op == ">":
                return left > right
            if op == "<":
                return left < right
            if op == ">=":
                return left >= right
            if op == "<=":
                return left <= right
        return left

    def unary() -> Any:
        if match_op("!"):
            return not unary()
        return postfix()

    # JS→Python method/property mapping for strings and lists
    _JS_STR_METHODS: dict[str, Any] = {
        "toUpperCase": lambda s: s.upper,
        "toLowerCase": lambda s: s.lower,
        "trim": lambda s: s.strip,
        "trimStart": lambda s: s.lstrip,
        "trimEnd": lambda s: s.rstrip,
        "startsWith": lambda s: s.startswith,
        "endsWith": lambda s: s.endswith,
        "includes": lambda s: lambda sub: sub in s,
        "indexOf": lambda s: lambda sub: s.find(sub),
        "slice": lambda s: lambda *a: s[int(a[0]):int(a[1])] if len(a) > 1 else s[int(a[0]):],
        "substring": lambda s: lambda *a: s[int(a[0]):int(a[1])] if len(a) > 1 else s[int(a[0]):],
        "replace": lambda s: lambda old, new: _js_replace(s, old, new, False),
        "replaceAll": lambda s: lambda old, new: _js_replace(s, old, new, True),
        "split": lambda s: lambda sep: s.split(sep),
        "charAt": lambda s: lambda i: s[int(i)] if 0 <= int(i) < len(s) else "",
        "repeat": lambda s: lambda n: s * int(n),
        "padStart": lambda s: lambda n, ch=" ": s.rjust(int(n), ch),
        "padEnd": lambda s: lambda n, ch=" ": s.ljust(int(n), ch),
    }
    _JS_LIST_METHODS: dict[str, Any] = {
        "join": lambda a: lambda sep: sep.join(str(x) for x in a),
        "includes": lambda a: lambda v: v in a,
        "indexOf": lambda a: lambda v: a.index(v) if v in a else -1,
        "slice": lambda a: lambda *args: a[int(args[0]):int(args[1])] if len(args) > 1 else a[int(args[0]):],
        "map": lambda a: lambda fn: [fn(x) for x in a],
        "filter": lambda a: lambda fn: [x for x in a if fn(x)],
        "find": lambda a: lambda fn: next((x for x in a if fn(x)), None),
        "reverse": lambda a: lambda: list(reversed(a)),
        "flat": lambda a: lambda *_: [item for sub in a for item in (sub if isinstance(sub, list) else [sub])],
        "concat": lambda a: lambda *args: a + [item for arg in args for item in (arg if isinstance(arg, list) else [arg])],
    }

    def postfix() -> Any:
        val = primary()
        while True:
            if match_op("."):
                ident = adv()
                if val is not None:
                    name = ident["v"]
                    # Handle .length for strings and lists
                    if name == "length" and isinstance(val, (str, list)):
                        val = len(val)
                    elif isinstance(val, str) and name in _JS_STR_METHODS:
                        val = _JS_STR_METHODS[name](val)
                    elif isinstance(val, list) and name in _JS_LIST_METHODS:
                        val = _JS_LIST_METHODS[name](val)
                    elif isinstance(val, dict):
                        val = val.get(name)
                    else:
                        prop = getattr(val, name, None)
                        val = prop
                else:
                    val = None
            elif match_op("["):
                idx = ternary()
                match_op("]")
                if val is not None:
                    if isinstance(val, (list, tuple)):
                        val = val[int(idx)] if isinstance(idx, (int, float)) else None
                    elif isinstance(val, dict):
                        val = val.get(str(idx))
                    else:
                        val = None
                else:
                    val = None
            elif match_op("("):
                args: list[Any] = []
                if not match_op(")"):
                    args.append(ternary())
                    while match_op(","):
                        args.append(ternary())
                    match_op(")")
                val = val(*args) if callable(val) else None
            else:
                break
        return val

    def primary() -> Any:
        t = peek()
        if t is None:
            return None
        if t["t"] == "n":
            adv()
            return t["v"]
        if t["t"] == "s":
            adv()
            return t["v"]
        if t["t"] == "r":
            adv()
            return t["v"]  # compiled regex
        if t["t"] == "i":
            adv()
            v = t["v"]
            if v == "true":
                return True
            if v == "false":
                return False
            if v == "null":
                return None
            if v == "undefined":
                return None
            return ctx.get(v)
        if t["t"] == "o" and t["v"] == "(":
            adv()
            v = ternary()
            match_op(")")
            return v
        # Array literal
        if t["t"] == "o" and t["v"] == "[":
            adv()
            arr: list[Any] = []
            if not match_op("]"):
                arr.append(ternary())
                while match_op(","):
                    arr.append(ternary())
                match_op("]")
            return arr
        # Object literal
        if t["t"] == "o" and t["v"] == "{":
            adv()
            obj: dict[str, Any] = {}
            if not match_op("}"):
                k = adv()
                match_op(":")
                obj[str(k["v"])] = ternary()
                while match_op(","):
                    k2 = adv()
                    match_op(":")
                    obj[str(k2["v"])] = ternary()
                match_op("}")
            return obj
        return None

    return ternary()


def interpolate(text: str, ctx: dict[str, Any]) -> str:
    if "{{" not in text:
        return text

    last_close = text.rfind("}}")
    trim_tail = last_close != -1 and text[last_close + 2 :].strip() == ""

    def replacer(m: re.Match) -> str:
        v = eval_expr(m.group(1).strip(), ctx)
        if v is None:
            return ""
        if isinstance(v, bool):
            return ""
        return str(v)

    out = re.sub(r"\{\{(.+?)\}\}", replacer, text)
    if trim_tail:
        out = out.rstrip()
    return out


def interpolate_pre(text: str, ctx: dict[str, Any]) -> str:
    if "{{" not in text:
        return text

    def replacer(m: re.Match) -> str:
        v = eval_expr(m.group(1).strip(), ctx)
        if v is None:
            return ""
        if isinstance(v, bool):
            return ""
        return str(v)

    return re.sub(r"\{\{(.+?)\}\}", replacer, text)


def eval_condition(expr: str, ctx: dict[str, Any]) -> bool:
    m = re.match(r"^\{\{(.+?)\}\}$", expr)
    if m:
        return bool(eval_expr(m.group(1).strip(), ctx))
    return bool(eval_expr(expr, ctx))


def escapes(text: str) -> str:
    text = text.replace("#quot;", '"')
    text = text.replace("#apos;", "'")
    text = text.replace("#lt;", "<")
    text = text.replace("#gt;", ">")
    text = text.replace("#lbrace;", "{")
    text = text.replace("#rbrace;", "}")
    text = text.replace("#amp;", "\x00")
    text = text.replace("#hash;", "#")
    # XML entities
    text = text.replace("&quot;", '"')
    text = text.replace("&apos;", "'")
    text = text.replace("&lt;", "<")
    text = text.replace("&gt;", ">")
    text = text.replace("&amp;", "")
    text = text.replace("\x00", "&")
    return text
